File: app/src/main/java/com/whiteside/insta/ui/ChatActivity.java
package com.whiteside.insta.ui;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.whiteside.insta.databinding.ActivityChatBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import Controller.MessagesRecyclerViewAdapter;
import Model.Message;
import Model.Profile;

import static android.content.ContentValues.TAG;

public class ChatActivity extends AppCompatActivity {

    private List<Message> messages;

    private String UID;
    private MessagesRecyclerViewAdapter adapter;

    private FirebaseAuth auth;
    private DatabaseReference dRef;
    private FirebaseFirestore fStore;
    private ActivityChatBinding bind;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        bind = ActivityChatBinding.inflate(getLayoutInflater());
        setContentView(bind.getRoot());

        setSupportActionBar(bind.appBar);

        UID = getIntent().getExtras().getString("UID");

        auth = FirebaseAuth.getInstance();
        fStore = FirebaseFirestore.getInstance();
        dRef = FirebaseDatabase.getInstance().getReference();

        bind.recyclerView.setHasFixedSize(true);
        bind.recyclerView.setLayoutManager(new LinearLayoutManager(this));

        messages = new ArrayList();
        adapter = new MessagesRecyclerViewAdapter(messages, this);
        bind.recyclerView.setAdapter(adapter);


        fStore.collection("Users")
                .document(UID)
                .addSnapshotListener(new EventListener<DocumentSnapshot>() {
                    @Override
                    public void onEvent(@Nullable DocumentSnapshot ds, @Nullable FirebaseFirestoreException e) {
                        if (e != null) {
                            Log.e(TAG, e.getStackTrace().toString());
                        } else if (ds.exists()) {
                            Profile profile = ds.toObject(Profile.class);

                            Bitmap bitmap = BitmapFactory.decodeByteArray(profile.getProfilePic().toBytes()
                                    , 0, profile.getProfilePic().toBytes().length);
                            bind.profileImage.setImageBitmap(bitmap);
                            bind.name.setText(profile.getName());
                        }
                    }
                });

        dRef.child("Users")
                .child(auth.getUid())
                .child("Chats")
                .child(UID)
                .addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                        int c = 1;
                        for (DataSnapshot s : snapshot.getChildren()) {
                            if (c > messages.size()) {
                                Message m = s.getValue(Message.class);
                                messages.add(m);
                                adapter.notifyDataSetChanged();
                                bind.recyclerView.smoothScrollToPosition(messages.size());
                            }
                            c++;
                        }
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {

                    }
                });
    }

    public void sendClicked(View view) {
        String messageText = bind.messageSend.getText().toString();
        if (!messageText.equals("")) {
            Message m = new Message();
            m.setFrom(auth.getUid());
            m.setTo(UID);
            m.setSeconds(Timestamp.now().getSeconds());
            m.setMessage(bind.messageSend.getText().toString());

            sendMessage(m);
        }
    }

    protected Boolean sendMessage(Message m) {
        DatabaseReference dbReferences = FirebaseDatabase.getInstance().getReference();
        FirebaseAuth auth = FirebaseAuth.getInstance();

        String uid = Objects.requireNonNull(auth.getUid());
        dbReferences = dbReferences
                .child("Users")
                .child(uid)
                .child("Chats")
                .child(m.getTo())
                .push();

        dbReferences.setValue(m);

        dbReferences
                .child("Users")
                .child(m.getTo())
                .child("Chats")
                .child(uid)
                .push()
                .setValue(m);

        return true;
    }
}

File: app/src/main/java/com/whiteside/insta/ChatActivity.kt
package com.whiteside.insta

import Controller.MessagesRecyclerViewAdapter
import Model.Message
import com.whiteside.insta.ui.edit_profile.Profile
import android.content.ContentValues
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.insta.databinding.ActivityChatBinding
import java.util.*
import kotlin.collections.ArrayList

class ChatActivity : AppCompatActivity() {
    private val messages: ArrayList<Message> = ArrayList()
    private var UID: String? = null
    private var adapter: MessagesRecyclerViewAdapter? = null
    private var auth: FirebaseAuth? = null
    private var dRef: DatabaseReference? = null
    private var fStore: FirebaseFirestore? = null
    private var bind: ActivityChatBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind = ActivityChatBinding.inflate(layoutInflater)
        setContentView(bind!!.root)
        setSupportActionBar(bind!!.appBar)
        UID = intent.extras!!.getString("UID")
        auth = FirebaseAuth.getInstance()
        fStore = FirebaseFirestore.getInstance()
        dRef = FirebaseDatabase.getInstance().reference
        bind!!.recyclerView.setHasFixedSize(true)
        bind!!.recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = MessagesRecyclerViewAdapter(messages, this)
        bind!!.recyclerView.adapter = adapter
        fStore!!.collection("Users")
                .document(UID!!)
                .addSnapshotListener { ds, e ->
                    if (e != null) {
                        Log.e(ContentValues.TAG, e.stackTrace.toString())
                    } else if (ds!!.exists()) {
                        val profile = ds.toObject(Profile::class.java)
                        val bitmap = BitmapFactory.decodeByteArray(profile!!.profilePic!!.toBytes(), 0, profile.profilePic!!.toBytes().size)
                        bind!!.profileImage.setImageBitmap(bitmap)
                        bind!!.name.text = profile.name
                    }
                }
        dRef!!.child("Users")
                .child(auth!!.uid!!)
                .child("Chats")
                .child(UID!!)
                .addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        var c = 1
                        for (s in snapshot.children) {
                            if (c > messages.size) {
                                val m = s.getValue(Message::class.java)
                                messages.add(m!!)
                                adapter!!.notifyDataSetChanged()
                                bind!!.recyclerView.smoothScrollToPosition(messages.size)
                            }
                            c++
                        }
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
    }

    fun sendClicked(view: View?) {
        val messageText = bind!!.messageSend.text.toString()
        if (messageText != "") {
            val m = Message()
            m.from = auth!!.uid
            m.to = UID
            m.seconds = Timestamp.now().seconds
            m.message = bind!!.messageSend.text.toString()
            sendMessage(m)
        }
    }

    protected fun sendMessage(m: Message): Boolean {
        var dbReferences = FirebaseDatabase.getInstance().reference
        val auth = FirebaseAuth.getInstance()
        val uid = Objects.requireNonNull(auth.uid)
        dbReferences = dbReferences
                .child("Users")
                .child(uid!!)
                .child("Chats")
                .child(m.to!!)
                .push()
        dbReferences.setValue(m)
        dbReferences
                .child("Users")
                .child(m.to!!)
                .child("Chats")
                .child(uid)
                .push()
                .setValue(m)
        return true
    }
}